internal/proxy/add.sh

```diff
@@ -0,0 +1,63 @@
+#!/bin/bash
+# add.sh adds new dependencies to the Go Cloud module proxy. This script
+# should be run from the root of the repository.
+
+set -o pipefail
+
+function log() {
+	echo "add.sh: $@" 1>&2
+}
+
+function die() {
+	log "$@"
+	exit 1
+}
+
+if ! [[ -d .git ]]; then die 'this script must be run from the root of the repository'; fi
+
+# This approach was suggested by:
+# https://github.com/go-modules-by-example/index/tree/master/012_modvendor
+
+log "creating temp dir where we'll create a module download cache"
+tgp="$(mktemp -d)"
+if [[ $? != 0 ]]; then die 'failed'; fi
+
+temp=$(mktemp)
+if [[ $? != 0 ]]; then die 'failed to create temp file for verbose command output'; fi
+
+# Copy current module cache into temporary directory as basis.
+# Periodically, someone on the project should go through this process without
+# running this step to prune unused dependencies.
+log 'making temp download dir'
+if ! mkdir -p "$tgp/pkg/mod/cache/download"; then die 'failed'; fi
+log "downloading current module cache, logging to $temp"
+if ! gsutil -m rsync -r gs://go-cloud-modules "$tgp/pkg/mod/cache/download" > $temp 2>&1; then die 'failed'; fi
```

Comment:
Here and elsewhere: I think I'm okay with having the logs be directed away on the happy path, but it is much easier for an end-user to see the logs on failure rather than having to open a file that isn't being cleaned up.

Reply:
I added traps to clean up the temp dir and modvendor but I left the temp dir cleanup for the end so users can manually inspect it in case of failure. I agree that it's easier in a sense if that file gets automatically sent to the terminal where the script is being run, but that also means it can be harder to see the context of what the script was trying to do when it failed.